monitor/pubsubmon/pubsubmon.go

```diff
@@ -0,0 +1,320 @@
+// Package pubsubmon implements a PeerMonitor component for IPFS Cluster that
+// uses PubSub to send and receive metrics.
+package pubsubmon
+
+import (
+	"bytes"
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ipfs/ipfs-cluster/api"
+	"github.com/ipfs/ipfs-cluster/monitor/util"
+
+	rpc "github.com/hsanjuan/go-libp2p-gorpc"
+	logging "github.com/ipfs/go-log"
+	floodsub "github.com/libp2p/go-floodsub"
+	host "github.com/libp2p/go-libp2p-host"
+	peer "github.com/libp2p/go-libp2p-peer"
+	msgpack "github.com/multiformats/go-multicodec/msgpack"
+)
+
+var logger = logging.Logger("monitor")
+
+// PubsubTopic specifies the topic used to publish Cluster metrics.
+var PubsubTopic = "pubsubmon"
+
+// AlertChannelCap specifies how much buffer the alerts channel has.
+var AlertChannelCap = 256
+
+// DefaultWindowCap sets the amount of metrics to store per peer.
+var DefaultWindowCap = 25
+
+var msgpackHandle = msgpack.DefaultMsgpackHandle()
+
+// Monitor is a component in charge of monitoring peers, logging
+// metrics and detecting failures
+type Monitor struct {
+	ctx       context.Context
+	cancel    func()
+	rpcClient *rpc.Client
+	rpcReady  chan struct{}
+
+	host         host.Host
+	pubsub       *floodsub.PubSub
+	subscription *floodsub.Subscription
+
+	metrics    util.Metrics
+	metricsMux sync.RWMutex
+	windowCap  int
+
+	checker *util.MetricsChecker
+	alerts  chan api.Alert
+
+	config *Config
+
+	shutdownLock sync.Mutex
+	shutdown     bool
+	wg           sync.WaitGroup
+}
+
+// New creates a new monitor. It receives the window capacity
```

Comment:
This docstring needs updating.

Reply:
Should the `WindowCap` be configurable?

Reply:
I don't see why, at this stage. I would want more clever failure detection (degradation), which may prompt to make it configurable. But out of scope here.